runtime/lua/vim/lsp/_snippet.lua

```diff
@@ -0,0 +1,421 @@
+  P.take_until({ '/' }, { '\\' }),
+  S.slash,
+  P.many(P.any(S.format, S.text({ '$', '/' }, { '\\' }))),
+  S.slash,
+  P.pattern('[ig]+')
+), function(values)
+  return {
+    type = NodeType.TRANSFORM,
+    pattern = values[2].raw,
+    format = values[4],
+    option = values[6],
+  }
+end)
+
+S.tabstop = P.any(
+  P.map(P.seq(S.dollar, S.int), function(values)
+    return {
+      type = NodeType.TABSTOP,
+      tabstop = values[2],
+    }
+  end),
+  P.map(P.seq(S.dollar, S.open, S.int, S.close), function(values)
+    return {
+      type = NodeType.TABSTOP,
+      tabstop = values[3],
+    }
+  end),
+  P.map(P.seq(S.dollar, S.open, S.int, S.transform, S.close), function(values)
+    return {
+      type = NodeType.TABSTOP,
+      tabstop = values[3],
+      transform = values[4]
+    }
+  end)
+)
+
+S.placeholder = P.any(
+  P.map(P.seq(S.dollar, S.open, S.int, S.colon, P.many(P.any(S.toplevel, S.text({ '$', '}' }, { '\\' }))), S.close), function(values)
+    return {
+      type = NodeType.PLACEHOLDER,
+      tabstop = values[3],
+      children = values[5],
+    }
+  end)
+)
+
+S.choice = P.map(P.seq(
+  S.dollar,
+  S.open,
+  S.int,
+  S.pipe,
+  P.many(
+    P.map(P.seq(S.text({ ',', '|' }), P.opt(S.comma)), function(values)
+      return values[1].esc
+    end)
+  ),
+  S.pipe,
+  S.close
+), function(values)
+  return {
+    type = NodeType.CHOICE,
+    tabstop = values[3],
+    items = values[5],
+  }
+end)
+
+S.variable = P.any(
+  P.map(P.seq(S.dollar, S.var), function(values)
+    return {
+      type = NodeType.VARIABLE,
+      name = values[2],
+    }
+  end),
+  P.map(P.seq(S.dollar, S.open, S.var, S.close), function(values)
+    return {
+      type = NodeType.VARIABLE,
+      name = values[3],
+    }
+  end),
+  P.map(P.seq(S.dollar, S.open, S.var, S.transform, S.close), function(values)
+    return {
+      type = NodeType.VARIABLE,
+      name = values[3],
+      transform = values[4],
+    }
+  end),
+  P.map(P.seq(S.dollar, S.open, S.var, S.colon, P.many(P.any(S.toplevel, S.text({ '$', '}' }, { '\\' }))), S.close), function(values)
+    return {
+      type = NodeType.VARIABLE,
+      name = values[3],
+      children = values[5],
+    }
+  end)
+)
+
+S.snippet = P.map(P.many(P.any(S.toplevel, S.text({ '$' }, { '}', '\\' }))), function(values)
+  return {
+    type = NodeType.SNIPPET,
+    children = values,
+  }
+end)
+
+local M = {}
+
+---@types table<string, number>
+---The snippet node type enum
+M.NodeType = NodeType
+
+---Parse snippet string and returns the AST
+---@param input string
+---@return table
+M.parse = function(input)
+  local result = S.snippet(input, 1)
+  if not result.parsed then
+    error('snippet parsing failed.')
+  end
+  return result.value
+end
+
+---Convert snippet nodes to string.
+---@param node table The result of M.parse
+---@param opts table|nil
+---  - with_placeholder boolean
+---  - resolve_variable fun(name: string): string
+---@return string
+M.to_string = function(node, opts)
```

Comment:
why introduce this idiosyncratic to_string function instead of overriding `__tostring` ? https://stackoverflow.com/a/35327975/152142

Reply:
how does that conflict with my suggestion?

Reply:
You are right, inspect, does not trigger   `__tostring`, so it's safe to use it there.

Reply:
If I implement `__tostring`, is it for each node?

Anyway, I will push the `__tostring` version.

Reply:
> If I implement `__tostring`, is it for each node?

isn't `parse()` already return some sort of custom object?

Reply:
I think `M.__tostring` is confusing because the function does not stringify `M` itself.

Reply:
No, you have to make a table like `{__tostring =  function ....}`  and then make that table the metatable of each node.

Reply:
I've pushed the `__tostring` version. What do you think about it?
If I misunderstood it, I will revert (remove by force-push) it.

Reply:
In my personal opinion, I like the `M.to_string` way because it allows customizing the format via the arguments in the future.

Reply:
> I've pushed the `__tostring` version. What do you think about it?

at a glance, looks good!



> In my personal opinion, I like the `M.to_string` way because it allows customizing the format via the arguments in the future.

we could introduce `fmt()` as a convention.

Reply:
@justinmk did you want to move to the fmt convention before merging this, or are we ok to merge as is?

Reply:
> @justinmk did you want to move to the fmt convention before merging this, or are we ok to merge as is?

i'd say wait until we need it. `__tostring()` seems like it should be the first instinct.

Reply:
Ok, cool, so I think the only blocking issue is the function style https://github.com/neovim/neovim/pull/15579#discussion_r704533106?